enterprise/server/scheduling/scheduler_server/scheduler_server.go

```diff
@@ -1802,6 +1759,52 @@ func getNonPreferredSchedulingDelay(cmd *repb.Command) time.Duration {
 	return d
 }
 
+func successfulReservation(node *executionNode, enqueueStart time.Time) string {
+	return fmt.Sprintf("%s [%s]", node.String(), time.Since(enqueueStart).String())
+}
+
+func (s *SchedulerServer) enqueue(ctx context.Context, node *executionNode, request *scpb.EnqueueTaskReservationRequest, opts enqueueTaskReservationOpts) bool {
+	if opts.scheduleOnConnectedExecutors {
+		return enqueueOnConnectedExecutor(ctx, node, request)
+	} else {
+		return s.enqueueOnRemoteExecutor(ctx, node, request)
+	}
+}
+
+func enqueueOnConnectedExecutor(ctx context.Context, node *executionNode, request *scpb.EnqueueTaskReservationRequest) bool {
+	if node.handle == nil {
+		log.CtxErrorf(ctx, "nil handle for a local executor %q", node.GetExecutorID())
+		return false
+	}
+	_, err := node.handle.EnqueueTaskReservation(ctx, request)
+	if err != nil {
+		log.Infof("failed to enqueue task on connected executor: %s", err)
+	}
+	return err == nil
+}
+
+func (s *SchedulerServer) enqueueOnRemoteExecutor(ctx context.Context, node *executionNode, request *scpb.EnqueueTaskReservationRequest) bool {
+	if node.schedulerHostPort == "" {
+		log.CtxErrorf(ctx, "node %q has no scheduler host:port set", node.GetExecutorID())
+		return false
+	}
+
+	schedulerClient, err := s.schedulerClientCache.get(node.schedulerHostPort)
+	if err != nil {
+		log.CtxWarningf(ctx, "Could not get SchedulerClient for %q: %s", node.schedulerHostPort, err)
+		return false
+	}
+	rpcCtx, cancel := context.WithTimeout(ctx, schedulerEnqueueTaskReservationTimeout)
+	_, err = schedulerClient.EnqueueTaskReservation(rpcCtx, request)
+	cancel()
+	if err != nil {
+		log.CtxWarningf(ctx, "EnqueueTaskReservation via scheduler target %q failed: %s", node.schedulerHostPort, err)
+		time.Sleep(schedulerEnqueueTaskReservationFailureSleep)
```

Comment:
Should we move this sleep outside the func and sleep whenever it returns false? (I think it would help us avoid getting into a hot CPU loop if something is borked.)

Alternatively we could return `(bool, error)` and abort the scheduling loop if a "fatal" error is returned. (non-fatal errors would return `(false, nil)`)

Reply:
That seems like a reasonable goal. I went ahead and did the `(bool, error)` approach because that let me preserve the existing behavior as-is, always sleeping on false would change the behavior by introducing a new sleep in a few cases which I'd like to avoid in this PR. LMK what you think.